scene/3d/collision_object_3d.cpp

```diff
@@ -171,57 +165,48 @@ void CollisionObject3D::_update_pickable() {
 	}
 }
 
+void CollisionObject3D::_clear_debug_shapes() {
+	for (Map<uint32_t, ShapeData>::Element *E = shapes.front(); E; E = E->next()) {
+		ShapeData &shapedata = E->get();
+		ShapeData::ShapeBase *shapes = shapedata.shapes.ptrw();
+		for (int i = 0; i < shapedata.shapes.size(); i++) {
+			ShapeData::ShapeBase &s = shapes[i];
+			if (s.debug_shape.is_valid()) {
+				RS::get_singleton()->free(s.debug_shape);
+			}
+		}
+	}
+}
+
 void CollisionObject3D::_update_debug_shapes() {
 	for (Set<uint32_t>::Element *shapedata_idx = debug_shapes_to_update.front(); shapedata_idx; shapedata_idx = shapedata_idx->next()) {
 		if (shapes.has(shapedata_idx->get())) {
 			ShapeData &shapedata = shapes[shapedata_idx->get()];
 			ShapeData::ShapeBase *shapes = shapedata.shapes.ptrw();
 			for (int i = 0; i < shapedata.shapes.size(); i++) {
 				ShapeData::ShapeBase &s = shapes[i];
-				if (s.debug_shape) {
-					s.debug_shape->queue_delete();
-					s.debug_shape = nullptr;
-					--debug_shape_count;
+				if (s.debug_shape.is_null()) {
+					s.debug_shape = RS::get_singleton()->instance_create();
+					RS::get_singleton()->instance_set_scenario(s.debug_shape, get_world_3d()->get_scenario());
 				}
 				if (s.shape.is_null() || shapedata.disabled) {
+					RS::get_singleton()->free(s.debug_shape);
+					s.debug_shape = RID();
 					continue;
 				}
-
 				Ref<Mesh> mesh = s.shape->get_debug_mesh();
-				MeshInstance3D *mi = memnew(MeshInstance3D);
-				mi->set_transform(shapedata.xform);
-				mi->set_mesh(mesh);
-				add_child(mi);
-				mi->force_update_transform();
-				s.debug_shape = mi;
-				++debug_shape_count;
+				RS::get_singleton()->instance_set_base(s.debug_shape, mesh->get_rid());
+				RS::get_singleton()->instance_set_transform(s.debug_shape, get_global_transform() * shapedata.xform);
```

Comment:
Why not keeping the same logic as before here? It seems a bit easier to handle:
1. Delete the debug shape if valid
2. Continue if there's no valid shape to draw
3. Create the debug shape instance

Reply:
The old approach was copied from CollisionShape3D. I took the opportunity to make it more efficient, so the instance is created *only* if it doesn't exist, instead of creating it every time.